src/WebJobs.Script.WebHost/Diagnostics/ConsoleWriter.cs

```diff
@@ -0,0 +1,169 @@
+                    var writeTask = _consoleBuffer.Writer.WaitToWriteAsync().AsTask();
+                    if (writeTask.WaitAsync(_consoleBufferTimeout).Result == false)
+                    {
+                        // The buffer has been completed and does not allow further writes.
+                        // Currently this should not be possible, but the safest thing to do is just write directly to the console.
+                        Console.WriteLine(evt);
+                        break;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                // Most likely a task cancellation exception from the timeout expiring, but regardless we use the handler
+                // dump the raw exception and write the event to the console directly
+                _exceptionhandler(ex);
+                Console.WriteLine(evt);
+            }
+        }
+
+        internal async Task ProcessConsoleBufferAsync()
+        {
+            try
+            {
+                await foreach (var line in _consoleBuffer.Reader.ReadAllAsync())
+                {
+                    Console.WriteLine(line);
+                }
+            }
+            catch (Exception ex)
+            {
+                // Not sure what can fail here, but there is nothing else monitoring this task so just log the exception
+                _exceptionhandler(ex);
+            }
+            finally
+            {
+                // if this has failed for any reason, complete the buffer so that everything falls back to console
+                _consoleBuffer.Writer.Complete();
+            }
+        }
+
+        internal async Task ProcessConsoleBufferBatchedAsync()
+        {
+            try
+            {
+                var builder = new StringBuilder();
+
+                while (true)
+                {
+                    if (await _consoleBuffer.Reader.WaitToReadAsync() == false)
+                    {
+                        // The buffer has been completed and does not allow further reads.
+                        // Currently this should not be possible, but safest thing to do is break out of the loop.
+                        break;
+                    }
+
+                    if (_consoleBuffer.Reader.TryRead(out string line1))
+                    {
+                        // Can we synchronously read multiple lines?
+                        // If yes, use the string builder to batch them together into a single write
+                        // If no, just write the single line without using the builder;
+                        if (_consoleBuffer.Reader.TryRead(out string line2))
+                        {
+                            builder.AppendLine(line1);
+                            builder.AppendLine(line2);
+
+                            while (_consoleBuffer.Reader.TryRead(out string nextLine))
```

Comment:
Should this have a cap per-outer-loop? The build could get very, very large under load if we get here.

Reply:
I upped this to a much higher value - the limit of how many lines we'll write to the builder before we flush and clear is now the same as the buffer size.

Reply:
Further investigation revealed the tradeoff in this setting - you can shift the latency distribution with this value. For example a high value of 8000 will keep most percentiles low, but push the P99 out quite high:

Max lines - 8000:
```
  Latency Distribution
     50%     5.53ms
     75%     6.53ms
     90%     7.53ms
     95%     9.01ms
     99%   692.64ms
```     
While a lower value of 800 will push out the P90 and P95, but keep the P99 under control.

Max lines - 800:
```
  Latency Distribution
     50%     6.00ms
     75%     9.77ms
     90%    70.05ms
     95%    72.17ms
     99%    75.81ms
```

PR has been updated to prefer the latter.